mavros_extras/src/plugins/odom.cpp

```diff
@@ -0,0 +1,156 @@
+/**
+ * @brief Odometry plugin
+ * @file odom.cpp
+ * @author James Goppert <james.goppert8@gmail.com>
+ *
+ * @addtogroup plugin
+ * @{
+ */
+/*
+ * Copyright 2017 James Goppert
+ *
+ * This file is part of the mavros package and subject to the license terms
+ * in the top-level LICENSE file of the mavros repository.
+ * https://github.com/mavlink/mavros/tree/master/LICENSE.md
+ */
+
+#include <mavros/mavros_plugin.h>
+#include <eigen_conversions/eigen_msg.h>
+
+#include <nav_msgs/Odometry.h>
+
+namespace mavros {
+namespace extra_plugins {
+using mavlink::common::MAV_ESTIMATOR_TYPE;
+/**
+ * @brief Odometry plugin
+ *
+ * Send odometry info
+ * to FCU position and attitude estimators
+ */
+class OdometryPlugin : public plugin::PluginBase {
+public:
+	OdometryPlugin() : PluginBase(),
+		odom_nh("~odometry"),
+		estimator_type(3)
+	{ }
+
+	void initialize(UAS &uas_)
+	{
+		PluginBase::initialize(uas_);
+
+		// general params
+		odom_nh.param("estimator_type", estimator_type, 3);	// defaulted to VIO type
+
+		ROS_DEBUG_STREAM_NAMED("odom", "Odometry: estimator type: " << utils::to_string_enum<MAV_ESTIMATOR_TYPE>(estimator_type));
+
+		// subscribers
+		odom_sub = odom_nh.subscribe("odom", 10, &OdometryPlugin::odom_cb, this);
+	}
+
+	Subscriptions get_subscriptions()
+	{
+		return { /* Rx disabled */ };
+	}
+
+private:
+	ros::NodeHandle odom_nh;
+	ros::Subscriber odom_sub;
+
+	int estimator_type;
+
+	/* -*- callbacks -*- */
+
+	void odom_cb(const nav_msgs::Odometry::ConstPtr &odom)
+	{
+		Eigen::Affine3d tr;
+		Eigen::Vector3d lin_vel_enu;
+		Eigen::Vector3d ang_vel_enu;
+		tf::poseMsgToEigen(odom->pose.pose, tr);
+		tf::vectorMsgToEigen(odom->twist.twist.linear, lin_vel_enu);
+		tf::vectorMsgToEigen(odom->twist.twist.angular, ang_vel_enu);
+
+		Eigen::Quaterniond enu_orientation;
+		tf::quaternionMsgToEigen(m_uas->get_attitude_orientation(), enu_orientation);
+
+		// body frame rotations must be aware of current attitude of the vehicle
+		auto q_ned_current = ftf::transform_orientation_enu_ned(
+					ftf::transform_orientation_baselink_aircraft(enu_orientation));
```

Comment:
After testing on simulation I'm mostly sure this is correct 'cause the transform is in body frame, which means it needs to follow the attitude of the body (aircraft) frame itself. With only applying a transform without making a rotation to the current attitude, I was having incorrect yaw rotations.
I'm also sure that this may be the src of some problems people are having with the mocap attitude having some weird rotations. When I make sure this works I'll probably assert of the transforms we have on the mocap plugin and check what are the variables that should be on world frame and what are the ones that need to be on body frame.